tools/nymvisor/src/cli/mod.rs

```diff
@@ -0,0 +1,157 @@
+            Commands::AddUpgrade(args) => add_upgrade::execute(args),
+            Commands::Config(args) => config::execute(args),
+        }
+    }
+}
+
+#[derive(Subcommand, Debug)]
+pub(crate) enum Commands {
+    /// Initialise a nymvisor instance with persistent Config.toml file.
+    Init(Box<init::Args>),
+
+    /// Run the associated daemon with the preconfigured settings.
+    Run(run::Args),
+
+    /// Show build information of this binary
+    BuildInfo(build_info::Args),
+
+    /// Show build information of the associated daemon
+    DaemonBuildInfo(daemon_build_info::Args),
+
+    /// Queues up another upgrade for the associated daemon
+    AddUpgrade(add_upgrade::Args),
+
+    /// Show configuration options being used by this instance of nymvisor
+    Config(config::Args),
+}
+
+fn open_config_file(env: &Env) -> Result<Config, NymvisorError> {
+    let config_load_location = if let Some(config_path) = &env.nymvisor_config_path {
+        config_path.clone()
+    } else if let Some(nymvisor_id) = &env.nymvisor_id {
+        // if no explicit path was provided in the environment, try to use the default one based on the nymvisor id
+        default_config_filepath(nymvisor_id)
+    } else {
+        // finally, if all else fails, see if this is a singleton -> if so try to load the only instance
+        try_get_singleton_nymvisor_config_path()?
+    };
+
+    debug!(
+        "attempting to load configuration file from {}",
+        config_load_location.display()
+    );
+
+    if let Ok(cfg) = Config::read_from_toml_file(&config_load_location) {
+        return Ok(cfg);
+    }
+
+    // we couldn't load it - try upgrading it from older revisions
+    try_upgrade_config(&config_load_location)?;
+
+    match Config::read_from_toml_file(&config_load_location) {
+        Ok(cfg) => Ok(cfg),
+        Err(source) => {
+            error!("Failed to load config from {}. Are you sure you have run `init` before? (Error was: {source})", config_load_location.display());
+            Err(NymvisorError::ConfigLoadFailure {
+                id: env.nymvisor_id.clone().unwrap_or("UNKNOWN".to_string()),
+                path: config_load_location,
+                source,
+            })
+        }
+    }
+}
+
+// attempt to get a path to nymvisor's config path if there is only a single instance
+pub(crate) fn try_get_singleton_nymvisor_config_path() -> Result<PathBuf, NymvisorError> {
+    let instances_dir = default_instances_directory();
+    let mut instances = instances_dir
+        .read_dir()
+        .map_err(|source| NymvisorError::InstancesReadFailure {
+            source,
+            path: instances_dir.clone(),
+        })?
+        .collect::<Vec<_>>();
+
+    if instances.len() != 1 {
+        return Err(NymvisorError::NotSingleton {
+            instances: instances.len(),
+        });
+    }
+
+    // safety: that unwrap is fine as we've just checked we have 1 entry in the vector
+    #[allow(clippy::unwrap_used)]
```

Comment:
Move it inside the else branch above instead?

Reply:
it'd have a slightly different meaning then. if I moved it to the `else` branch I'd be checking if I have **at least** once instance. meanwhile I want to check I have **exactly** one instance